CONFIG.md
# Configuration File Explanation

This document describes the purpose and usage of each section and setting in the provided configuration file. The file is used to manage the behavior of the tomato monitoring system.

## **[model]** 👾
| Setting              | Description                                                                                                      |
|----------------------|------------------------------------------------------------------------------------------------------------------|
| `size_model`         | Path to the model weights used for detecting the size or ripeness of tomatoes.                                   |
| `disease_model`      | Path to the model weights used for detecting tomato diseases.                                                    |
| `device`             | Acceleration device to be used when detecting. Default is `cpu`, but `cuda` or the index of the GPU can be used. |
| `size_confidence`    | Confidence threshold for the size/ripeness detection model. Default: `0.6`.                                      |
| `disease_confidence` | Confidence threshold for the disease detection model. Default: `0.8`.                                            |
| `size_ripened_count` | Minimum number of ripened tomatoes required to trigger an alert for gathering. Default: `10`.                    |
| `input`              | Path to the input video or stream file for processing. Default: `test_tomato.mp4`.                               |
| `output`             | Directory to save the processed results. Default: `test_results`.                                                |
| `detection_cooldown` | Time interval between consecutive detections (in seconds). Default: `25`.                                        |
| `use_minutes`        | Boolean to treat `detection_cooldown` as minutes when `True`. Default: `False`.                                  |
| `show_stream`        | Boolean indicating whether to display the video stream. Default: `True`.                                         |
| `save_images`        | Boolean to save detection images to the output folder. Default: `False`.                                         |

---

## **[chart]** 📊
| Setting     | Description                                                                 |
|-------------|-----------------------------------------------------------------------------|
| `output`    | Directory to save results used for generating charts. Default: `test_results`. |
| `subdir`    | Subdirectory within the output folder to save charts. Default: `charts`.    |
| `save`      | Boolean to enable or disable saving charts. Default: `False`.               |

---

## **[email]** 📧
| Setting      | Description                                                                 |
|--------------|-----------------------------------------------------------------------------|
| `smtp_server`| The SMTP server for sending emails. Default: `smtp.gmail.com`.              |
| `port`       | Port number for the SMTP server. Recommended: `587` for secure connections. |
| `sender`     | Email address sending the alerts.                                           |
| `password`   | Password for the sender's email account. **Remove before sharing code.**    |
| `receiver`   | Recipient email address for receiving alerts.                               |

---

##  **[database]** 📚
| Setting      | Description                                                                 |
|--------------|-----------------------------------------------------------------------------|
| `host`       | Database server address. Default: `localhost`.                              |
| `port`       | Port number for connecting to the database. Default: `3307`.                |
| `name`       | Name of the database to use. Default: `tomato`.                             |
| `username`   | Username for accessing the database. Default: `root`.                       |
| `password`   | Password for the database user. Default: `123456`.                          |
| `use_database`| Boolean to enable or disable using the database. Default: `False`.         |

---

## **[ftp]** 📤
| Setting     | Description                                                                 |
|-------------|-----------------------------------------------------------------------------|
| `server`    | Address of the FTP server. Default: `0.0.0.0`.                              |
| `user`      | Username for the FTP server. Default: `user`.                               |
| `port`      | Port number for the FTP connection. Default: `2121`.                        |
| `password`  | Password for the FTP user. Default: `123456`.                               |
| `use_ftp`   | Boolean to enable or disable using the FTP server. Default: `False`.        |


config.py
import configparser

try:
    config_file = configparser.ConfigParser()
    config_file.read('config.ini')
except Exception as e:
    print(e)
    print('No config.ini file was found!')


class Config:
    """
    Class for easy config file management
    Note: You can use it to make config object for each config selection
    """

    def __init__(self, section: str):
        self.section = section

    def get(self, key: str, fallback=None, type: str = ''):
        """
        Used for efficient retrieving of data
        :param key: Name of the key to be retrieved
        :param fallback: Default value in case key does not exist or is not set
        :param type: (Optional) Type specification for types different than 'str'
        :return: The called variable from config file. Depending on the type it can be 'str', 'bool', 'int', 'float'
        """
        if type == 'bool':
            return config_file.getboolean(self.section, key, fallback=fallback)
        if type == 'int':
            return config_file.getint(self.section, key, fallback=fallback)
        if type == 'float':
            return config_file.getfloat(self.section, key, fallback=fallback)
        else:
            return config_file.get(self.section, key, fallback=fallback)


class TomatoModelConfig:
    def __init__(self):
        config = Config('model')

        self.device = config.get('device', 'cpu')
        try:
            self.device = int(self.device)
            print('GPU acceleration is enabled!')
        except:
            print(f'GPU acceleration is disabled!')

        self.size_tomato_model_path: str = config.get('size_model', 'size.pt')
        self.disease_tomato_model_path: str = config.get('disease_model', 'disease.pt')

        self.size_tomato_confidence: float = config.get('size_confidence', 0.8, type='float')
        self.disease_tomato_confidence: float = config.get('disease_confidence', 0.8, type='float')

        self.size_tomato_ripened_count: int = config.get('size_ripened_count', 10, type='int')

        self.source: str = config.get('input', 'test_inputs/')
        self.output_folder: str = config.get('output', 'test_results/')

        self.cooldown: int = config.get('detection_cooldown', 10, type='int')
        self.use_minutes: str = config.get('use_minutes', True, type='bool')
        if self.use_minutes:
            self.cooldown: int = self.cooldown * 60

        self.show_stream: bool = config.get('show_stream', True, type='bool')
        self.save_images: bool = config.get('save_images', True, type='bool')


class ChartConfig:
    def __init__(self):
        config = Config('chart')

        self.output: str = config.get('output', 'test_results/')
        self.subdir: str = config.get('subdir', 'chart/')
        self.save: bool = config.get('save', False, type='bool')


class EmailConfig:
    def __init__(self):
        config = Config('email')

        self.smtp_server: str = config.get('smtp_server', 'smtp.gmail.com')
        self.port: int = config.get('port', 25, type='int')
        self.sender: str = config.get('sender', 'me@gmail.com')
        self.receiver: str = config.get('receiver', 'you@gmail.com')
        self.password: str = config.get('password', '123456')


class DatabaseConfig:
    def __init__(self):
        config = Config('database')

        self.use_database: bool = config.get('use_database', False, type='bool')
        if not self.use_database:
            print('Database writing is disabled!')
            return

        self.username: str = config.get('username', 'root')
        self.host: str = config.get('host', 'localhost')
        self.port: int = config.get('port', 3306, type='int')
        self.database_name: str = config.get('name', 'tomato')
        self.password: str = config.get('password', '123456')

        self.URL: str = f'mysql+pymysql://{self.username}:{self.password}@{self.host}/{self.database_name}'


class FTPConfig:
    def __init__(self):
        config = Config('ftp')

        self.use_ftp: bool = config.get('use_ftp', True, type='bool')
        if not self.use_ftp:
            print('Ftp writing is disabled!')
            return

        self.server: str = config.get('server', 'ftp.example.com')
        self.port: int = config.get('port', 2121, type='int')
        self.user: str = config.get('user', 'some username')
        self.password: str = config.get('password', 'password')


tomato_model_config = TomatoModelConfig()
chart_config = ChartConfig()
email_config = EmailConfig()
database_config = DatabaseConfig()
ftp_config = FTPConfig()

if __name__ == '__main__':
    print(tomato_model_config.save_images)
    print(tomato_model_config.show_stream)
    print(tomato_model_config.cooldown)

    print(chart_config.subdir)
